### PR TITLE
Rework files related to DashboardPicker to TS

### PR DESCRIPTION
**Pull request by uladzimirdev**
related to https://github.com/metabase/metabase/issues/29666

### Description

Basically this PR converts js and jsx code related to ItemPicker and Modal to TS.

requires rework of Form to be merged first 
https://github.com/metabase/metabase/pull/30418

### How to verify

typecheck and tests should pass

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30224)

### CODE REVIEW

File: frontend/src/metabase/query_builder/components/view/ViewHeader.jsx
Comment (uladzimirdev) [inline frontend/src/metabase/query_builder/components/view/ViewHeader.jsx:349]: removes annoying warning

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (uladzimirdev) [inline frontend/src/metabase/query_builder/components/QueryModals.tsx:105]: IDK, I found it more readable, but I'm fine with reverthing this

File: frontend/src/metabase/dashboard/containers/DashboardHeader.jsx
Comment (uladzimirdev) [inline frontend/src/metabase/dashboard/containers/DashboardHeader.jsx:100]: fixes annoying warnings in console

File: frontend/src/metabase/entities/containers/EntityCopyModal.tsx
Comment (ranquild) [inline frontend/src/metabase/entities/containers/EntityCopyModal.tsx:10]: string?

File: frontend/src/metabase/entities/containers/EntityCopyModal.tsx
Comment (ranquild) [inline frontend/src/metabase/entities/containers/EntityCopyModal.tsx:11]: <T>?

**Review by ranquild — APPROVED** [review]

File: frontend/src/metabase/containers/DashboardPicker.tsx
Comment (kulyk) [inline frontend/src/metabase/containers/DashboardPicker.tsx:10]: Not sure these comments carry a lot of meaning here

File: frontend/src/metabase/containers/AddToDashSelectDashModal.tsx
Comment (kulyk) [inline frontend/src/metabase/containers/AddToDashSelectDashModal.tsx:2]: Could we remove both of these "disable" comments? Prop types should be fine now and order should be easy to fix

File: frontend/src/metabase/containers/AddToDashSelectDashModal.tsx
Comment (kulyk) [inline frontend/src/metabase/containers/AddToDashSelectDashModal.tsx:59]: Do we need this comment?

File: frontend/src/metabase/containers/AddToDashSelectDashModal.tsx
Comment (kulyk) [inline frontend/src/metabase/containers/AddToDashSelectDashModal.tsx:76]: I think we don't need a `@ts-expect-error` here. Canonical root collection's ID is `null`, so the types should respect that

File: frontend/src/metabase/containers/AddToDashSelectDashModal.tsx
Comment (kulyk) [inline frontend/src/metabase/containers/AddToDashSelectDashModal.tsx:93]: Can we remove this comment?

File: frontend/src/metabase/containers/SaveQuestionModal.tsx
Comment (kulyk) [reply frontend/src/metabase/containers/SaveQuestionModal.tsx:14]: Looks good to me; what makes you think something's off here? I'd only ask to move the import up, we prefer keeping local imports at the very end of the list

File: frontend/src/metabase/containers/SaveQuestionModal.tsx
Comment (kulyk) [inline frontend/src/metabase/containers/SaveQuestionModal.tsx:47]: What can we do to avoid the expect-error comment here?

File: frontend/src/metabase/containers/SaveQuestionModal.tsx
Comment (kulyk) [inline frontend/src/metabase/containers/SaveQuestionModal.tsx:57]: Please use `CollectionId` type from `metabase-types/api`

File: frontend/src/metabase/containers/SaveQuestionModal.tsx
Comment (kulyk) [inline frontend/src/metabase/containers/SaveQuestionModal.tsx:205]: Could we specify props?

**Review by kulyk — COMMENTED** [review]
Looks good, I'm mostly concerted about a handful of `@ts-expect-error` comments this PR introduces

File: frontend/src/metabase/entities/containers/EntityCopyModal.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/entities/containers/EntityCopyModal.tsx:11]: maybe, but I'm not sure as I didn't find any example of similar type I believe

File: frontend/src/metabase/containers/DashboardPicker.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/containers/DashboardPicker.tsx:10]: those are leftovers from the propTypes, I'll remove them :)

File: frontend/src/metabase/containers/AddToDashSelectDashModal.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/containers/AddToDashSelectDashModal.tsx:76]: the beauty of `ts-expect-error` is that it will fail as soon as there is no ts error, I even have a PR about it https://github.com/metabase/metabase/pull/29796

probably I can create RFC explaining this, should I?

File: frontend/src/metabase/containers/AddToDashSelectDashModal.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/containers/AddToDashSelectDashModal.tsx:93]: it's for reviewer

**Comment by uladzimirdev** [comment]
> Looks good, I'm mostly concerted about a handful of `@ts-expect-error` comments this PR introduces

@kulyk do you think comments are misleading?

File: frontend/src/metabase/containers/AddToDashSelectDashModal.tsx
Comment (kulyk) [reply frontend/src/metabase/containers/AddToDashSelectDashModal.tsx:76]: Yeah, I think if there's a way to avoid using it, we should go for it — collection ID can be either a number or a null (also could be "root" when we need an ID-collection map), so there should be a way to properly type this

Re: RFC — ofc!

File: frontend/src/metabase/containers/AddToDashSelectDashModal.tsx
Comment (kulyk) [reply frontend/src/metabase/containers/AddToDashSelectDashModal.tsx:93]: I'd encourage you to use GitHub comments for that, so these comments don't stick around in the codebase for more that we actually need

File: frontend/src/metabase/containers/SaveQuestionModal.tsx
Comment (kulyk) [reply frontend/src/metabase/containers/SaveQuestionModal.tsx:205]: If you mean a type is coming from `metabase-types/types`, it should be fine to use it when typing old components. Ideally, we'd want to use `metabase-types/api`  for new code and refactor the old code to use them as well

File: frontend/src/metabase/containers/AddToDashSelectDashModal.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/containers/AddToDashSelectDashModal.tsx:93]: you're right, but at the same time I can forget about this change and it's easy to overlook it

**Comment by kulyk** [comment]
> do you think comments are misleading?

@uladzimirdev re `@ts-expect-error` comments, I believe they could be really helpful as a last resort of some kind
If there's a way to type things and avoid the comment, I think we should prefer that and I believe we have some room to make things that way

**Comment by uladzimirdev** [comment]
please take a look one more time, main change is [here](https://github.com/metabase/metabase/pull/30224/commits/b0c7d940c3d4c7ca9356a60c5e45ff9cd054dcb9), it's related to the form rework

**Comment by uladzimirdev** [comment]
@ranquild @kulyk 
please take a look one more time, main change is [here](https://github.com/metabase/metabase/pull/30224/commits/b0c7d940c3d4c7ca9356a60c5e45ff9cd054dcb9), it's related to the form rework

**Comment by uladzimirdev** [comment]
rework of the form can be a different PR to the master

**Review by ranquild — COMMENTED** [review]
@uladzimirdev could you merge the 2 base PRs with refactoring & fixes and update this one?

> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/30224#-:-NU1SYjYEar29uSCywPJ:bmd2vcy)* status: 0 of 19 files reviewed, 15 unresolved discussions (waiting on @kulyk and @uladzimirdev)

File: frontend/src/metabase/entities/containers/EntityCopyModal.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/entities/containers/EntityCopyModal.tsx:11]: I left it as any

**Comment by uladzimirdev** [comment]
> > do you think comments are misleading?
> 
> @uladzimirdev re `@ts-expect-error` comments, I believe they could be really helpful as a last resort of some kind If there's a way to type things and avoid the comment, I think we should prefer that and I believe we have some room to make things that way

@kulyk I fixed all the ts-expect-error statements you complained above, form related ts-expect-error are covered in scope of https://github.com/metabase/metabase/pull/30418

File: frontend/src/metabase/containers/AddToDashSelectDashModal.tsx
Comment (kulyk) [inline frontend/src/metabase/containers/AddToDashSelectDashModal.tsx:58]: Should we maybe coerce `dashboardId` to string/number (what we need there)? Like that:

```js
dashboards[String(dashboardId)]
```

File: frontend/src/metabase/containers/SaveQuestionModal.tsx
Comment (kulyk) [inline frontend/src/metabase/containers/SaveQuestionModal.tsx:91]: Is it possible to refactor this so we can call `isOriginalQuestionNotNullable` only once?

File: frontend/src/metabase/containers/SaveQuestionModal.unit.spec.tsx
Comment (kulyk) [inline frontend/src/metabase/containers/SaveQuestionModal.unit.spec.tsx:102]: I think we prefer casting queries to `StructuredQuery` in cases like that:

```ts
const query = originalQuestion.query() as StructuredQuery;
return query.breakout(...).question().markDirty(); 
```

File: frontend/src/metabase/containers/SaveQuestionModal.unit.spec.tsx
Comment (kulyk) [inline frontend/src/metabase/containers/SaveQuestionModal.unit.spec.tsx:201]: Should we maybe refactor this test to use `expect(...).toHaveBeenCalledOnce()` instead?

File: frontend/src/metabase/entities/containers/EntityCopyModal.tsx
Comment (kulyk) [reply frontend/src/metabase/entities/containers/EntityCopyModal.tsx:11]: Can we maybe use `unknown` instead? `any` could be too loose

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/QueryModals.tsx:44]: Can we use a real type here?

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/QueryModals.tsx:47]: Is it really a number?

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/QueryModals.tsx:50]: Can we use a real config type here? Maybe export it [from `metabase/query_builder/actions/core/updateQuestion`](https://github.com/metabase/metabase/blob/7b034eb869d9aa0958e228f423c0627415dfb58c/frontend/src/metabase/query_builder/actions/core/updateQuestion.ts#L100)?

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/QueryModals.tsx:51]: Can we use [the `QueryBuilderMode` type from `metabase-types/store/qb`](https://github.com/metabase/metabase/blob/7b034eb869d9aa0958e228f423c0627415dfb58c/frontend/src/metabase-types/store/qb.ts#L7) here?

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/QueryModals.tsx:52]: Not sure `originalQuestion` is always provided

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/QueryModals.tsx:55]: Not sure if these types match the current state of the query builder:

1. [`onCreate` accepts a single argument](https://github.com/metabase/metabase/blob/7b034eb869d9aa0958e228f423c0627415dfb58c/frontend/src/metabase/query_builder/containers/QueryBuilder.jsx#L267)
2. [`onSave` accepts two arguments, but the second one should be an object](https://github.com/metabase/metabase/blob/7b034eb869d9aa0958e228f423c0627415dfb58c/frontend/src/metabase/query_builder/containers/QueryBuilder.jsx#L277)

I could be wrong here, though — QB prop drilling is 🫠

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/QueryModals.tsx:63]: I believe the first argument should be `Partial<Card>` (we should make the ID required tho)

File: frontend/src/metabase/containers/AddToDashSelectDashModal.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/containers/AddToDashSelectDashModal.tsx:58]: I'm fine with either approach

Comment (uladzimirdev) [reply frontend/src/metabase/containers/AddToDashSelectDashModal.tsx:58]: actually dashboardId can be null|undefined as well, so it's better to have a check if it exists

File: frontend/src/metabase/containers/SaveQuestionModal.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/containers/SaveQuestionModal.tsx:91]: yes, possible, I thought having the previous structure would simplify review

File: frontend/src/metabase/containers/SaveQuestionModal.unit.spec.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/containers/SaveQuestionModal.unit.spec.tsx:102]: image
I don't see `breakout` in the api folder at all. Do you think we miss it?

Comment (uladzimirdev) [reply frontend/src/metabase/containers/SaveQuestionModal.unit.spec.tsx:102]: found a solution 🙌

File: frontend/src/metabase/containers/SaveQuestionModal.unit.spec.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/containers/SaveQuestionModal.unit.spec.tsx:201]: we have checks for the argument later in the code, so `call` is useful anyway

```
      const newQuestion = call[0];
      expect(newQuestion.id()).toBeUndefined();
      expect(newQuestion.displayName()).toBe(EXPECTED_SUGGESTED_NAME);
      expect(newQuestion.description()).toBe(null);
      expect(newQuestion.collectionId()).toBe(null);
```

I doubt I'll be able to replace it with `expect().toHaveBeenCalledWith(expect.objectContaining({}))`, so maybe it's better to keep it as it is

File: frontend/src/metabase/entities/containers/EntityCopyModal.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/entities/containers/EntityCopyModal.tsx:11]: I believe unknown is not the best type, as we'll need to cast it somehwhere anyway. What actually types can be used? I can think of Question,Dashboard,Collection?

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/query_builder/components/QueryModals.tsx:47]: it's drilled down as a number, so seems yes

image.png

File: frontend/src/metabase/containers/SaveQuestionModal.unit.spec.tsx
Comment (kulyk) [reply frontend/src/metabase/containers/SaveQuestionModal.unit.spec.tsx:201]: We can use both ways at the same time, but fwiw I'm fine keeping it as is

File: frontend/src/metabase/entities/containers/EntityCopyModal.tsx
Comment (kulyk) [reply frontend/src/metabase/entities/containers/EntityCopyModal.tsx:11]: We can give it a shot unless it turns into a mess; hopefully we'll be able to provide more type-safety with entity hooks.

> What actually types can be used? I can think of Question,Dashboard,Collection?

I think there should be a type in `metabase-types/api` per each possible entity. Think we should use `Card` instead of `Question` tho

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/query_builder/components/QueryModals.tsx:44]: can you please help me with types here? I didn't find actual places from where to grab the type

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/query_builder/components/QueryModals.tsx:52]: nice catch

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/query_builder/components/QueryModals.tsx:55]: maybe you're right here, I don't fully understand prop drilling as well, but from what I see, we do not touch directly lines you provided here, I tried to record explanation here https://www.loom.com/share/82b44c7bf89a4fcb8c6ab8eb7fb0216f

second half of the video can be skipped 😄

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/query_builder/components/QueryModals.tsx]: done!

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/query_builder/components/QueryModals.tsx]: done

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/query_builder/components/QueryModals.tsx]: fixed, added `| null`

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/query_builder/components/QueryModals.tsx]: another great catch, changed it to `Pick<Card, 'id'>`

File: frontend/src/metabase/containers/SaveQuestionModal.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/containers/SaveQuestionModal.tsx]: reworked this, two separate methods seems more readable here

File: frontend/src/metabase/containers/AddToDashSelectDashModal.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/containers/AddToDashSelectDashModal.tsx]: fixed a type, also added additional `if`, so no coercing required

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/query_builder/components/QueryModals.tsx:49]: yes, it's a number of the event and we pass it further. Not the best name IMO

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (kulyk) [reply frontend/src/metabase/query_builder/components/QueryModals.tsx:46]: Yeah, seems like we don't have types for alerts yet. #30457 should add them; feel free to cherry-pick the commits here or we can wait until its approved and merged

File: frontend/src/metabase/query_builder/components/QueryModals.tsx
Comment (kulyk) [reply frontend/src/metabase/query_builder/components/QueryModals.tsx:57]: I suppose these signatures are just outdated; I assume the boolean argument was used to trigger query re-runs. And now, `QueryModals` either passes `false` as a second arg for `onCreate` and `onSave`, or doesn't pass the flag at all. Given that `onCreate` only accepts a single argument and `onSave` expects an optional object, both these signatures have no effect.

Also, doesn't seem like there's any intermediate component that could wrap `onCreate` and `onSave` into anything that accepts a boolean flag. `QueryModals` is only rendered by QB's `View` component ([here](https://github.com/metabase/metabase/blob/7b034eb869d9aa0958e228f423c0627415dfb58c/frontend/src/metabase/query_builder/components/view/View.jsx#L514) and [here](https://github.com/metabase/metabase/blob/7b034eb869d9aa0958e228f423c0627415dfb58c/frontend/src/metabase/query_builder/components/view/View.jsx#L468)). In both cases, `View` just spreads all its props for the modals component. And `View` is a [direct child of the `QueryBuilder` component](https://github.com/metabase/metabase/blob/7b034eb869d9aa0958e228f423c0627415dfb58c/frontend/src/metabase/query_builder/containers/QueryBuilder.jsx#L402) that defines both `onCreate` and `onSave`

Comment (uladzimirdev) [reply frontend/src/metabase/query_builder/components/QueryModals.tsx:57]: that's a great knowledge sharing 😄 thanks for so deep digging into code

**Review by kulyk — APPROVED** [review]
Looks good, thanks for addressing the comments!

File: frontend/src/metabase/entities/containers/EntityCopyModal.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/entities/containers/EntityCopyModal.tsx:11]: I'm leaving it as it is for now as it should be generic most likely, but I met also different problem that Question from metabase-lib relies on metabase-types which is deprecated
https://github.com/metabase/metabase/blob/master/frontend/src/metabase-lib/Question.ts#L38

and it leads ot type error even using generics becuase of differences between Card from API and Card from metabase-types/types, which leads to further TS errors.

Probably likely it worth having an issue for this refactoring

Comment (uladzimirdev) [reply frontend/src/metabase/entities/containers/EntityCopyModal.tsx:11]: https://github.com/metabase/metabase/issues/30471

**Review by ranquild — APPROVED** [review]

**Comment by uladzimirdev** [comment]
thanks @kulyk for deep and valuable feedback and help on making this PR mergeable 🙇

**Pull request merged**
